File: app/src/main/java/art/coded/wireframe/work/WorkViewModel.java
package art.coded.wireframe.work;

import android.app.Application;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.ViewModel;
import androidx.work.Constraints;
import androidx.work.OneTimeWorkRequest;
import androidx.work.WorkInfo;
import androidx.work.WorkManager;

import java.util.List;

public class WorkViewModel extends ViewModel {

    WorkManager mWorkManager;
    LiveData<List<WorkInfo>> mWorkInfo;

    public WorkViewModel(Application application) {

        mWorkManager = WorkManager.getInstance(application);
        mWorkInfo = mWorkManager.getWorkInfosByTagLiveData(WorkConstants.DEFAULT_WORK_TAG);
    }

    LiveData<List<WorkInfo>> getWorkInfo() { return mWorkInfo; }

    void applyWork() {

//        WorkContinuation workContinuation = mWorkManager.beginUniqueWork(
//                WorkConstants.DEFAULT_WORK_NAME,
//                ExistingWorkPolicy.REPLACE,
//                OneTimeWorkRequest.from(DifferentWorker.class)
//        ); workContinuation.enqueue();

        Constraints constraints = new Constraints.Builder()
                // set constraints for OneTimeWorkRequest
                .build();

        OneTimeWorkRequest workRequest = new OneTimeWorkRequest.Builder(DefaultWorker.class)
                .setConstraints(constraints)
                .addTag(WorkConstants.DEFAULT_WORK_TAG)
                .build();

//        workContinuation = workContinuation.then(workRequest);
//         chain additional requests
//        workContinuation.enqueue();

        mWorkManager.enqueue(workRequest);
    }

    void cancelWork() {
        mWorkManager.cancelUniqueWork(WorkConstants.DEFAULT_WORK_TAG);
    }
}

File: app/src/ktx/java/art/coded/wireframe/viewmodel/WorkViewModel.kt
package art.coded.wireframe.viewmodel

import androidx.lifecycle.LiveData
import art.coded.wireframe.view.work.WorkConstants
import art.coded.wireframe.view.work.DefaultWorker
import android.util.Log
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import androidx.work.*

private val LOG_TAG = WorkViewModel::class.java.simpleName

class WorkViewModel(private val workManager: WorkManager) : ViewModel() {

    val actionPending = MutableLiveData<Boolean>()

    fun getInfo(): LiveData<List<WorkInfo>> {
        actionPending.postValue(true)
        val workInfo: LiveData<List<WorkInfo>> = workManager.getWorkInfosByTagLiveData(WorkConstants.DEFAULT_WORK_TAG)
        actionPending.postValue(false)
        return workInfo
    }

    fun applyWork() {
        actionPending.postValue(true)
        val constraints = Constraints.Builder() // set constraints for OneTimeWorkRequest
            .build()
        val workRequest = OneTimeWorkRequest.Builder(DefaultWorker::class.java)
            .setConstraints(constraints)
            .addTag(WorkConstants.DEFAULT_WORK_TAG)
            .build()
        var workContinuation = workManager.beginUniqueWork(
            WorkConstants.DEFAULT_WORK_NAME,
            ExistingWorkPolicy.REPLACE,
            workRequest
        )
        workContinuation.enqueue()
        workContinuation = workContinuation.then(workRequest)
        //         chain additional requests
        workContinuation.enqueue()

//        workManager.enqueue(workRequest);
        Log.d(LOG_TAG, "enqueued")
        actionPending.postValue(false)
    }

    fun cancelWork() {
        actionPending.postValue(true)
        workManager.cancelUniqueWork(WorkConstants.DEFAULT_WORK_NAME)
        actionPending.postValue(false)
    }
}

class WorkViewModelFactory(private val workManager: WorkManager) : ViewModelProvider.Factory {
    override fun <T : ViewModel> create(modelClass: Class<T>): T {
        return WorkViewModel(workManager) as T
    }
}